### DISCUSSION TITLE
Орлан и ключ от пещеры

### DISCUSSION THREAD

**Discussion by dosinabox** [dosinabox/g2nr_unofficial_update#497] [General]
Или наличие взятого квеста про телепорты делать обязательным условием:Квест + кольцоИли Квес + членство в КВ.Тогда так же решится вопрос с тем что герой может попасть к магам воды через телепорты, раньше чем он придёт к ним с Ларесом. 

_Originally posted by @Kor-Angar in https://github.com/dosinabox/g2nr_unofficial_update/issues/41#issuecomment-545887055_

**Comment by dosinabox** [comment]
Вопрос интересный. Как ГГ вообще узнает, что Орлан связан с телепортами? В двух случаях:

1) об этом говорит Ларес:
```
AI_Output(self,other,"DIA_Addon_Lares_Teleportstation_09_04");	//Ну что ж, если тебе не страшно ими пользоваться, попроси трактирщика Орлана впустить тебя в его пещеру.
AI_Output(self,other,"DIA_Addon_Lares_Teleportstation_09_05");	//В ней спрятан один из телепортов магов Воды. Пещера находится неподалеку от его таверны.
```
2) ГГ сам догадывается спросить об этом Орлана, воспользовавшись телепортом около таверны

**Reply by Kor-Angar** [reply]
> ГГ сам догадывается спросить об этом Орлана, воспользовавшись телепортом около таверны

не так. герой после взятие квеста у мага воды, догадался спросить у Орлана так как знает что Орлан член КВ.

**Comment by dosinabox** [comment]
При этом для диалога с Ларесом "Тебе доводилось пользоваться телепортами?" (в котором он советует идти к Орлану) нужно **и** взять квест, **и** использовать любой телепорт. Наверное, нужно заменить условие на "взять квест **или** использовать любой телепорт".

**Reply by Kor-Angar** [reply]
всё верно. взять квест у Мага Воды и использовать телепорт который ведёт от магов воды.

**Reply by Kor-Angar** [reply]
> При этом для диалога с Ларесом "Тебе доводилось пользоваться телепортами?" (в котором он советует идти к Орлану) нужно и взять квест, и использовать любой телепорт. Наверное, нужно заменить условие на "взять квест или использовать любой телепорт".

**коммент от 2019 года**
Не, не. Наоборот правильное условие. И квест и телепорт.Там квест именно про то что ГГ их проверяет: работают они или нет.

**Comment by Kor-Angar** [comment]
Наверное даже вот так:
взять квест + использовать телепорты + показать аква кольцо 
или 
взять квест +  использовать телепорты + вступить в Кольцо
или
взять квест +  использовать телепорты + поговорить с Ларесом о телепортах

**Comment by Kor-Angar** [comment]
> Тогда так же решится вопрос с тем что герой может попасть к магам воды через телепорты, раньше чем он придёт к ним с Ларесом. 

Тут я оговорился. Перепутал телепорты. Герой может попасть к магом до того как придёт с Ларесом к ним

**Reply by Kor-Angar** [reply]
да. именно в этом суть. нельзя что бы герой проник к магам в первый через телепорт.

**Comment by Kor-Angar** [comment]
Там логика начинается с инфы о телепортах у Магов воды. Ключ от первого телепорта у мага воды, оттуда герой попадает в город.

```
instance DIA_Addon_Merdarion_Teleportstein(C_Info)
{
	npc = KDW_1405_Addon_Merdarion_NW;
	nr = 5;
	condition = DIA_Addon_Merdarion_Teleportstein_Condition;
	information = DIA_Addon_Merdarion_Teleportstein_Info;
	description = "Как быстрее всего добраться в город?";
};


func int DIA_Addon_Merdarion_Teleportstein_Condition()
{
//	if(Npc_KnowsInfo(other,DIA_Addon_Merdarion_Aufgabe) && (MIS_Addon_Lares_Ornament2Saturas == LOG_SUCCESS))
	if(Npc_KnowsInfo(other,DIA_Addon_Merdarion_Aufgabe))
	{
		return TRUE;
	};
};

func void DIA_Addon_Merdarion_Teleportstein_Info()
{
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_15_00");	//Как быстрее всего добраться в город?
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_01");	//Я тебе советую пойти тем же путем, что ты и пришел.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_02");	//(прикидывая) Но ты также можешь... Нет, это слишком опасно.
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_15_03");	//Выкладывай.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_04");	//Ладно. Строители этих залов передвигались своим способом.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_05");	//Насколько мы можем судить, они умели телепортироваться.
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_15_06");	//В этом нет ничего необычного.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_07");	//Телепортационные камни, которые мы нашли в Хоринисе, в самом деле, весьма необычны.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_08");	//Кажется, что телепортационный камень в этих залах может вести в город.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_06_09");	//Но ни у кого из нас не хватило смелости это проверить.
	Info_ClearChoices(DIA_Addon_Merdarion_Teleportstein);
	Info_AddChoice(DIA_Addon_Merdarion_Teleportstein,"Что если мне попробовать переместиться с его помощью?",DIA_Addon_Merdarion_Teleportstein_ich);
	Info_AddChoice(DIA_Addon_Merdarion_Teleportstein,"Где находится этот камень?",DIA_Addon_Merdarion_Teleportstein_wo);
};

func void DIA_Addon_Merdarion_Teleportstein_wo()
{
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_wo_15_00");	//Где находится этот камень?
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_wo_06_01");	//За дверью, в комнате с бассейном. Я запер его там.
	Info_AddChoice(DIA_Addon_Merdarion_Teleportstein,"Дай мне ключ. Я хочу посмотреть на него.",DIA_Addon_Merdarion_Teleportstein_key);
};

func void DIA_Addon_Merdarion_Teleportstein_ich()
{
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_ich_15_00");	//Что если мне попробовать переместиться с его помощью?
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_ich_06_01");	//Не знаю. Нам слишком мало об этом известно. Возможно, ты не выживешь.
};

func void DIA_Addon_Merdarion_Teleportstein_key()
{
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_key_15_00");	//Дай мне ключ. Я хочу посмотреть на него.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_key_06_01");	//(неуверенно) Правда? Но я не отвечаю за последствия.
	AI_Output(other,self,"DIA_Addon_Merdarion_Teleportstein_key_15_02");	//Давай ключ.
	AI_Output(self,other,"DIA_Addon_Merdarion_Teleportstein_key_06_03");	//Если ты так хочешь. Пожалуйста.
	CreateInvItems(self,ItKe_PortalTempelWalkthrough_Addon,1);
	B_GiveInvItems(self,other,ItKe_PortalTempelWalkthrough_Addon,1);
	SC_GotPortalTempelWalkthroughKey = TRUE;
	Log_CreateTopic(TOPIC_Addon_TeleportsNW,LOG_MISSION);
	Log_SetTopicStatus(TOPIC_Addon_TeleportsNW,LOG_Running);
	B_LogEntry(TOPIC_Addon_TeleportsNW,"Мердарион дал мне ключ от запертой двери. За ней находится телепорт, который перенесет меня обратно к городу.");
	Info_ClearChoices(DIA_Addon_Merdarion_Teleportstein);
};
```

Там он может расспросить Лареса, члена КВ и тот ему расскажет про Орлана. 

```
AI_Output(self,other,"DIA_Addon_Lares_Teleportstation_09_04");	//Ну что ж, если тебе не страшно ими пользоваться, попроси трактирщика Орлана впустить тебя в его пещеру.
AI_Output(self,other,"DIA_Addon_Lares_Teleportstation_09_05");	//В ней спрятан один из телепортов магов Воды. Пещера находится неподалеку от его таверны.
```

Или герой уже знает что Орлан член КВ и расспрашивает его сам.

********

условия на диалог с Орланом 

1. взять квест + использовать телепорты + показать аква кольцо 
или 
2. взять квест +  использовать телепорты + вступить в Кольцо
или
3. взять квест +  использовать телепорты + поговорить с Ларесом о телепортах

по сути если вариант 3, то герой всё равно должен показать Орлан кольцо, что бы дать понять что он в теме. т.е. может третий вариант и лишний, а может и нет, т.к. в игре есть момент когда ГГ уже вернул кольцо Ларесу, но еще не вступил в КВ.

**Comment by Kor-Angar** [comment]
есть разные моменты:
1) к Магам воды можно попасть через телепорт раньше чем Ларес отдаст орнамент
2) у Мердариона  есть реплика что герой к ним пришёл не через телепорт
3) у Лареса и Орлана разные условия на диалог про телепорт
4) Маг воды рассказывает про телепорт герою возле Таверны Мертвая гарпия, хотя герой о нем может уже знать

если добавить проверку Орлану на взятие квеста про телепорты, то
1) к Магам воды в первый раз нельзя будет попасть через телепорт. и это очень очень хорошо.
2) реплика Мердариона про то что герой пришёл пешком всегда будет актуальна
3) у Лареса и Орлана будут одинаковые условия на диалог про телепорт
4) герой не сможет узнать от Мердариона о том что телепорт возле Таверны после того как уже поговорит об этом с Орланом

**Comment by Kor-Angar** [comment]
> у Лареса и Орлана разные условия на диалог про телепорт

> у Лареса и Орлана будут одинаковые условия на диалог про телепорт

оказалось, что достаточно было заглянуть в код :-) https://github.com/Kor-Angar/G2_NotR_Kor_Angar/commit/3d6a979c1b2f25055990586d5b2c759492493324

**Comment by dosinabox** [comment]
> 3d6a979

У Лареса ничего не изменилось?)

**Comment by Kor-Angar** [comment]
всё так же :-)  с информационной справкой от Лареса проблем не наблюдалось. Это Орлан несознательно себя вёл :-)

**Comment by Kor-Angar** [comment]
хотя по хорошему вместо
(MIS_Addon_Lares_Ornament2Saturas == LOG_SUCCESS)
надо делать проверку на то взять ли квест про телепорты у обоих

**Comment by dosinabox** [comment]
И как одной емкой строкой описать это изменение, чтобы добавить его в общий список?

**Comment by Kor-Angar** [comment]
Герой не сможет попасть на раскопки  к магам воды через телепорт пока не вернет орнамент Сатурасу.
Или
Орлан не будет давать ключ от пещеры с телепортом герою, если герой  не получил задание о поисках телепорта.

**Comment by dosinabox** [comment]
> Орлан не будет давать ключ от пещеры с телепортом герою, если герой не получил задание о поисках телепорта.

Без этого задания ведь нельзя вступить в Кольцо Воды, так?
И какую проблему это решает?

**Comment by Kor-Angar** [comment]
> Без этого задания ведь нельзя вступить в Кольцо Воды, так?

можно.

> И какую проблему это решает?

нету проблем. мне кажется эта правка для неофициального обновления не подходит. она затрагивает небольшое кол-во игроков которые играют неправильно и в целом ниначто не влияет.

**Comment by Kor-Angar** [comment]
> Вопрос интересный. Как ГГ вообще узнает, что Орлан связан с телепортами? В двух случаях:

соответствующая запись появляется в диалоге после получения квеста.

**Comment by Kor-Angar** [comment]
я вчера поиграл. и в прошлом году играл и в позапрошлом. мне кажется самый правильный вариант на диалог с Орланом  это  **DIA_Addon_Orlan_Teleportstein**
1. Герой узнал об этом он Мердариона
```D
if(Npc_KnowsInfo(other,DIA_Addon_Merdarion_WHereOtherTeleports)
	{
		return TRUE;
	};
```
2. Герой узнал об это от Лареса
```D
if(Npc_KnowsInfo(other,DIA_Addon_Lares_Teleportstation)
	{
		return TRUE;
	};
```
3. герой взял квест Мердариона и показывал кольцо Орлану
```D
	if((SC_IsRanger == TRUE) || (Orlan_KnowsSCAsRanger == TRUE))
	{
		if  (SC_GotPortalTempelWalkthroughKey == TRUE)
		{
			return TRUE;
		};
	};
```

то что герой может получить у Орлана ключ к телепорту и через это заскачить к Магом Воды до того как сам или с Ларесом дойдёт до них это плохо.

**Reply by Kor-Angar** [reply]
```D
func int DIA_Addon_Orlan_Teleportstein_Condition()
{
	if  (SC_GotPortalTempelWalkthroughKey == false)
	{
			return false; // герой не взял квест на телепорты у Мердариона
	};

	if(Npc_KnowsInfo(other,DIA_Addon_Merdarion_WHereOtherTeleports)
	{
		return TRUE; // герой узнал инфу от Мердариона
	};

if(Npc_KnowsInfo(other,DIA_Addon_Lares_Teleportstation)
	{
		return TRUE; // герой узнал инфу от Лареса
	};

	if((SC_IsRanger == TRUE) || (Orlan_KnowsSCAsRanger == TRUE))
	{
		return TRUE; // герой догадался, так как знает что Онар член КВ
	};
};
```

**Reply by Kor-Angar** [reply]
а почему в программирование так любят уменьшать кол-во строк?

```D
func int DIA_Addon_Orlan_Teleportstein_Condition()
{
    if (C_Npc_KnowsInfo(DIA_Addon_Merdarion_WHereOtherTeleports)    { return SC_GotPortalTempelWalkthroughKey; };
    if (C_Npc_KnowsInfo(DIA_Addon_Lares_Teleportstation)            { return SC_GotPortalTempelWalkthroughKey; };
    if (SC_IsRanger == TRUE)                                        { return SC_GotPortalTempelWalkthroughKey; };
    if (Orlan_KnowsSCAsRanger == TRUE)                              { return SC_GotPortalTempelWalkthroughKey; };
};
```

**Reply by dosinabox** [reply]
Код окончательный? Можно просто скопировать его в проект и забыть об этой теме? Идет четвертый год, а понимание всё так же дается с трудом.

> а почему в программирование так любят уменьшать кол-во строк?

Я про такое не знаю.

> C_Npc_KnowsInfo

Что-то новенькое?

**Reply by Kor-Angar** [reply]
> Код окончательный? Можно просто скопировать его в проект и забыть об этой теме? 

я ещё не проверял.
```D
func int DIA_Addon_Orlan_Teleportstein_Condition()
{
    if  (SC_GotPortalTempelWalkthroughKey == false)
    {
            return false; // герой не взял квест на телепорты у Мердариона
    };

    if (Npc_KnowsInfo(hero,DIA_Addon_Merdarion_WHereOtherTeleports)
    {
        return TRUE; // герой узнал инфу от Мердариона
    };

    if (Npc_KnowsInfo(hero,DIA_Addon_Lares_Teleportstation)
    {
        return TRUE; // герой узнал инфу от Лареса
    };

    if ((SC_IsRanger == TRUE) || (Orlan_KnowsSCAsRanger == TRUE))
    {
        return TRUE; // герой догадался, так как знает что Онар член КВ
    };
};
```

> Идет четвертый год, а понимание всё так же дается с трудом.

герой может попасть к магам воды через телепорт до того как придёт туда с орнаментом Ватраса, так как в текущей версии Онар может дать герою ключ от телепорта. на мой взгляд это плохо.

![20230210020028_1](https://user-images.githubusercontent.com/49514764/218262264-590c873b-7129-4a46-9a27-d690dbfca406.jpg)
